In [144]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn import preprocessing
from sklearn.neighbors import KNeighborsClassifier

In [145]:
df2 = pd.read_csv('https://s3.us.cloud-object-storage.appdomain.cloud/cf-courses-data/CognitiveClass/DP0701EN/version-2/Data-Collisions.csv')
df=df2[['SEVERITYCODE','WEATHER','ROADCOND','LIGHTCOND']]

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3072: DtypeWarning: Columns (33) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [146]:
df2.SEVERITYCODE.value_counts()

1    136485
2     58188
Name: SEVERITYCODE, dtype: int64

In [147]:
#dropping all the null values from the data set
df=df.dropna()

In [148]:
#checking if null values still exists
df.isnull().sum().sum()

0

In [149]:
#Label Encoding- transforming the values into int64
from sklearn.preprocessing import LabelEncoder
labelencoder = LabelEncoder()
df['WEATHER'] = labelencoder.fit_transform(df['WEATHER'])
df['ROADCOND'] = labelencoder.fit_transform(df['ROADCOND'])  
df['LIGHTCOND'] = labelencoder.fit_transform(df['LIGHTCOND']) 
# df['WEATHER'].get_dummies()
df.head()

SEVERITYCODE  WEATHER  ROADCOND  LIGHTCOND
0             2        4         8          5
1             1        6         8          2
2             1        4         0          5
3             1        1         0          5
4             2        6         8          5

In [150]:
df['SEVERITYCODE'].value_counts()

1    132285
2     57052
Name: SEVERITYCODE, dtype: int64

In [23]:
from sklearn.utils import resample

In [12]:
#dont run this, this is just for practiving Hot Codes
#df=pd.get_dummies(df,columns=['WEATHER','ROADCOND','LIGHTCOND'])
#df.head()

SEVERITYCODE  WEATHER_0  WEATHER_1  WEATHER_2  WEATHER_3  WEATHER_4  \
0             2          0          0          0          0          1   
1             1          0          0          0          0          0   
2             1          0          0          0          0          1   
3             1          0          1          0          0          0   
4             2          0          0          0          0          0   

   WEATHER_5  WEATHER_6  WEATHER_7  WEATHER_8  ...  ROADCOND_8  LIGHTCOND_0  \
0          0          0          0          0  ...           1            0   
1          0          1          0          0  ...           1            0   
2          0          0          0          0  ...           0            0   
3          0          0          0          0  ...           0            0   
4          0          1          0          0  ...           1            0   

   LIGHTCOND_1  LIGHTCOND_2  LIGHTCOND_3  LIGHTCOND_4  LIGHTCOND_5  \
0            0            0            0            0            1   
1            0            1            0            0            0   
2            0            0            0            0            1   
3            0            0            0            0            1   
4            0            0            0            0            1   

   LIGHTCOND_6  LIGHTCOND_7  LIGHTCOND_8  
0            0            0            0  
1            0            0            0  
2            0            0            0  
3            0            0            0  
4            0            0            0  

[5 rows x 30 columns]

In [24]:
#df_maj=df[df.SEVERITYCODE==0]
#df_min=df[df.SEVERITYCODE==1]
#df_maj_dsample=resample(df_maj,replace=False,n_samples=58188,random_state=123)
#df_balanced=pd.concat([df_maj_dsample,df_min])
#df_balanced.SEVERITYCODE.value_counts()

ValueError: Cannot sample 58188 out of arrays with dim 0 when replace is False

In [151]:
df.dtypes

SEVERITYCODE    int64
WEATHER         int64
ROADCOND        int64
LIGHTCOND       int64
dtype: object

In [152]:
df['WEATHER'].value_counts()

1     111008
6      33117
4      27681
10     15039
9        901
3        824
2        569
8        113
0         55
7         25
5          5
Name: WEATHER, dtype: int64

In [153]:
df['ROADCOND'].value_counts()

0    124300
8     47417
7     15031
1      1206
5       999
3       131
6       115
4        74
2        64
Name: ROADCOND, dtype: int64

In [154]:
df['LIGHTCOND'].value_counts()

5    116077
2     48440
8     13456
6      5889
4      2502
0      1535
1      1192
7       235
3        11
Name: LIGHTCOND, dtype: int64

In [155]:
y=np.asarray(df['SEVERITYCODE'])
y[0:5]

array([2, 1, 1, 1, 2])

In [156]:
X=np.asarray(df[['WEATHER','ROADCOND','LIGHTCOND']])
X[0:5]


array([[4, 8, 5],
       [6, 8, 2],
       [4, 0, 5],
       [1, 0, 5],
       [6, 8, 5]])

In [143]:
#preprocessing
from sklearn import preprocessing
X=preprocessing.StandardScaler().fit(X).transform(X)
X[0:5]

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/utils/validation.py:595: DataConversionWarning: Data with input dtype int64 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/utils/validation.py:595: DataConversionWarning: Data with input dtype int64 was converted to float64 by StandardScaler.
  warnings.warn(msg, DataConversionWarning)


array([[ 0.32145781,  1.4789272 ,  0.35007251],
       [ 1.02224659,  1.4789272 , -1.40094051],
       [ 0.32145781, -0.71204349,  0.35007251],
       [-0.72972538, -0.71204349,  0.35007251],
       [ 1.02224659,  1.4789272 ,  0.35007251]])

In [157]:
#Train/Test Split
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3)
print('train set:', X_train.shape,y_train.shape)
print('Test set:',X_test.shape,y_test.shape)

train set: (132535, 3) (132535,)
Test set: (56802, 3) (56802,)


APPLYING THE MACHINE LEARNING TECHNIQUES:

                                                 1.  KNN(K-Nearest Neighbors)

In [158]:
k=20
knn=KNeighborsClassifier(n_neighbors = k).fit(X_train,y_train)
knn_y_pred = knn.predict(X_test)

In [159]:
#ACCURACY
acc = accuracy_score(y_test,knn_y_pred)

In [160]:
acc

0.6964191401711207

                                                   Results & Evaluation

In [161]:
from sklearn.metrics import jaccard_similarity_score
from sklearn.metrics import f1_score
trainScores_jaccard=jaccard_similarity_score(y_test, knn_y_pred)
trainScore_f1_score=f1_score(y_test, knn_y_pred, average='weighted')

In [162]:
trainScores_jaccard

0.6964191401711207

In [163]:
trainScore_f1_score

0.5718285326245788

Model Is most accurate when K=20: accuracy=0.6964, Jaccard=0.6964, f1_score=0.5718

                                                      2.  DECISION TREE

In [164]:
from sklearn.tree import DecisionTreeClassifier
dt=DecisionTreeClassifier(criterion="entropy",max_depth=7)
dt.fit(X_train,y_train)

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=7,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [165]:
dt_y_pred=dt.predict(X_test)
dt_y_pred

array([1, 1, 1, ..., 1, 1, 1])

In [166]:
#Accuracy
acc_decision = accuracy_score(y_test,dt_y_pred)
acc_decision

0.696331115101581

                                                 Results & Evaluation

In [167]:
from sklearn.metrics import jaccard_similarity_score
from sklearn.metrics import f1_score
jaccard1=jaccard_similarity_score(y_test, dt_y_pred)
f1_score1=f1_score(y_test, dt_y_pred, average='macro')

In [168]:
jaccard1

0.696331115101581

In [169]:
f1_score1

0.4105485595401346

Model is most accurate when depth=7: accuracy=0.6963, jaccard=0.6963,f1_score=0.4105

                                                  3.  LOGISTIC REGRESSION

In [170]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
logistic_dt=LogisticRegression(C=7, solver='liblinear').fit(X_train,y_train)

In [171]:
logistic_y_pred=logistic_dt.predict(X_test)
logistic_y_prob=logistic_dt.predict_proba(X_test)

In [172]:
logistic_y_pred

array([1, 1, 1, ..., 1, 1, 1])

In [173]:
logistic_y_prob

array([[0.67672847, 0.32327153],
       [0.67672847, 0.32327153],
       [0.67672847, 0.32327153],
       ...,
       [0.67672847, 0.32327153],
       [0.64995564, 0.35004436],
       [0.64615737, 0.35384263]])

In [174]:
#ACCURACY
acc_logistic = accuracy_score(y_test,logistic_y_pred)
acc_logistic

0.6963839301433048

                                                      Results & Evaluation

In [175]:
#jaccard and f1_Score
from sklearn.metrics import jaccard_similarity_score
from sklearn.metrics import f1_score
logistic_jaccard=jaccard_similarity_score(y_test, logistic_y_pred)
logistic_f1_score=f1_score(y_test, logistic_y_pred, average='macro')

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/sklearn/metrics/classification.py:1143: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


In [176]:
logistic_jaccard

0.6963839301433048

In [177]:
logistic_f1_score

0.4105108034620893

Model Is accurate when C=6: accuracy=0.6963, jaccard=0.6963, f1_score=0.4105

Combined Results:
                   
                   ML Model           ACCURACY   JACCARD SCORE  f1_score
                   
                   KNN                  0.6964       0.6964      0.5718 
                   
                   Decision tree        0.6963       0.6963      0.4105
                   
                   Logistic Regression  0.6963       0.6963      0.4105

So, From the results we can say that KNN is the best model to predict accident severity.